_posts/2019-01-30-year-of-the-dog.md

```diff
@@ -0,0 +1,199 @@
+
+We liked the experience so much we implemented a backend service for email templating
+in typescript and [wrote about
+it](https://tech.store2be.com/email/sendwithus/mjml/typescript/react/2018/06/14/email-templates-at-store2be-and-gdpr/).
+
+Another notable and much appreciated addition to our toolbox was
+[react-storybook](https://github.com/storybooks/storybook) for our shared UI
+components.
+
+## An unexpected journey - GraphQL
+
+In day to day frontend work however, architectural issues were still making our lives hard:
+
+1. Our RESTful API structure had not evolved much: most of our endpoints were
+   mapping directly to database table, with little additional logic.  With more
+   resources and more complex business logic, that meant **a lot of business
+   logic had to be implemented in the frontend** and merely validated by the
+   API.
+2. **State management and synchronization in the React apps was getting too
+   complex** - some operations and queries required several network calls and
+   dispatching multiple redux actions. Some critical operations were not
+   atomic, when they should have been.
+3. **Onboarding new developers to our redux-saga based abstractions** and all
+   the related conventions **proved challenging**, as documentation was not as
+   comprehensive as it should have been and old conventions coexisted with new
+   ones.
+4. With most of the data in our apps coming from the API, we felt that manually
+   writing Typescript interfaces for all these resources was a lot of error
+   prone manual work. We thought about generating them from our
+   [OpenAPI/Swagger](https://swagger.io/specification/) documentation, but the
+   experiment wasn't very conclusive. Coupled with the type information loss
+   from the redux-saga generators, we felt that we were **not realizing
+   Typescript's potential**.
+5. **Documentation was maintained manually as a Swagger specification**. This
+   was also a lot of error-prone, manual work.
+
+Faced with these problems, we decided to try and address them by progressively
+introducing a new GraphQL API, in the same Rails codebase as the existing
+RESTful API. Here is what we observed:
+
+1. The additional layer gave us more room and structure to **move business
+   logic back to the backend**.
+2. **Data fetching in the frontend is now a lot less complex**, with queries
+   mapping more or less 1-to-1 with views, and mutations to forms and buttons.
+   We are not using any redux-based abstraction for CRUD anymore, and the
+   overall amount of state management code has decreased. State is (or
+   appears) also more localized to where it belongs in the component tree. We
+   found this easier to reason about.
+3. The simplicity, but also the adherence to a well-adopted and documented
+   standard for data fetching made **onboarding easier**. There are a lot of
+   good GraphQL resources for frontend developers. In our experience picking up
```

Comment:
Would definitely list a couple of notable examples: even just the graphql.org/learn page was great.